manual/manual/cmds/instrumented-runtime.etex

```diff
@@ -0,0 +1,148 @@
+%HEVEA\cutname{instrumented-runtime.html}
+
+This chapter describes the OCaml instrumented runtime, a runtime variant
+allowing the collection of events and metrics.
+It aims to provide insight into a program's execution.
+
+\section{s:instr-runtime-overview}{Overview}
+
+Once compiled and linked with the instrumented runtime, any OCaml program
+can generate {\em trace files} that can then be read
+and analyzed by users in order to understand specific runtime behaviors.
+
+{\em Common Trace Format} is the general purpose binary
+format used to encode traces.
+A complete trace is comprised of:
+\begin{itemize}
+\item a {\em metadata file}, part of the OCaml distribution
+\item and a {\em trace file}, generated by the runtime\
+  in the program being traced.
+\end{itemize}
+
+For more information on {\em Common Trace Format}, see the website at
+\ifouthtml
+\ahref{https://diamon.org/ctf/}{https://diamon.org/ctf/}.
+\else
+{\tt https://diamon.org/ctf/}
+\fi
+
+\section{s:instr-runtime-enabling}{Enabling runtime instrumentation}
+
+Compiling a program and linking it with the instrumented runtime,
+can be done with {\tt -runtime-variant} flag {\tt i}:
+
+\begin{verbatim}
+       ocamlopt -runtime-variant i program.ml -o program
+\end{verbatim}
+
+The resulting program can then be traced by running it with the environment
+variable {\tt OCAML\_EVENTLOG\_ENABLED}:
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+The program will execute and generate a trace file in the
+current directory.
+
+\subsection{ss:instr-runtime-prefix}{Trace filename}
+
+The default trace filename is {\tt caml-\{pid\}-eventlog}, where {\tt \{pid\}}
+is the process identifier of the traced program.
+
+This filename can also be specified using the
+{\tt OCAML\_EVENTLOG\_FILE} environment variable.
+The given path will be suffixed with {\tt .pid.eventlog}.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_FILE=/tmp/some/prefix OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+In this example, the trace will be available at path
+{\tt /tmp/some/prefix.pid.eventlog}.
+
+\subsection{ss:instr-runtime-pause}{Pausing and resuming tracing}
+Primitives are available to control events collection at runtime.
+
+{\tt OCAML\_EVENTLOG\_ENABLED} can be set to the {\tt p} flag in order
+to {\em pause} events collection after the program has started.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=p ./program
+\end{verbatim}
+
+The program will have to resume events collection explicitly.
+This can be done via {\tt Gc.eventlog_resume} and
+{\tt Gc.eventlog_pause}, available in the {\stdmoduleref{Gc}} module.
+
+\subsection{ss:instr-runtime-limitations}{Limitations}
+
+The instrumented runtime does not support the {\tt fork} system call.
+A child process forked from an instrumented program will not be traced.
+
+The instrumented runtime aims to provide insight into the runtime's execution
+while maintaining a low overhead.
```

Comment:
Could you give a number of observed overhead on some interesting benchmark you did? Just "low" does not mean much to me, I would rather see "low (from X% to Y% in our tests)" to get an idea of what is your idea of "low".

Reply:
I think the new additions are is very useful, thanks! It gives user an idea of what to expect.